File: leetcode/RangeSumQuery2D-Immutable/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2020 by iamslash 

# 1 1 1 1
# 1 1 1 1
# 1 1 1 1
# 1 1 1 1

# 1 2  3  4
# 2 4  6  8
# 3 6  9 12
# 4 8 12 16
from typing import List

# 112ms 59.84% 17MB 6.24%
# grid sum
# O(HW) O(1)
class NumMatrix:
    def __init__(self, M: List[List[int]]):
        self.ps = M
        for y in range(len(M)):
            for x in range(len(M[0])):
                self.ps[y][x] += self.ps[y-1][x] if y > 0 else 0
                self.ps[y][x] += self.ps[y][x-1] if x > 0 else 0
                self.ps[y][x] -= self.ps[y-1][x-1] if y > 0 and x > 0 else 0
    def sumRegion(self, y1: int, x1: int, y2: int, x2: int) -> int:
        ans  = self.ps[y2][x2]
        ans -= self.ps[y1-1][x2] if y1 > 0 else 0
        ans -= self.ps[y2][x1-1] if x1 > 0 else 0
        ans += self.ps[y1-1][x1-1] if y1 > 0 and x1 > 0 else 0
        return ans

if __name__ == "__main__":
    main()

File: leetcode/RangeSumQuery2D-Immutable/a.go
/* Copyright (C) 2020 by iamslash */

package main

import "fmt"

// Wrong Answer:
// ["NumMatrix","sumRegion","sumRegion","sumRegion"]
// [[[[3,0,1,4,2],[5,6,3,2,1],[1,2,0,1,5],[4,1,0,1,7],[1,0,3,0,5]]],[2,1,4,3],[1,1,2,2],[1,2,2,4]] 
type NumMatrix struct {
  gs [][]int
}
func Constructor(M [][]int) NumMatrix {
  h, w := len(M), len(M[0])
  for y := 0; y < h; y++ {
    for x := 0; x < w; x++ {
      if y > 0 {
        M[y][x] += M[y-1][x]
      }
      if x > 0 {
        M[y][x] += M[y][x-1]
      }
      if y > 0 && x > 0 {
        M[y][x] -= M[y-1][x-1]
      }
    }
  }
  return NumMatrix{
    gs: M,
  }
}
func (this *NumMatrix) SumRegion(y1 int, x1 int, y2 int, x2 int) int {
  ans := this.gs[y2][x2]
  if y1 > 0 {
    ans -= this.gs[y1-1][x2]
  }
  if x1 > 0 {
    ans -= this.gs[y1][x1-1]
  }
  if y1 > 0 && x1 > 0 {
    ans += this.gs[y1-1][x1-1]
  }
  return ans
}
